### PR TITLE
sql,security: support storing SCRAM hashes in ALTER/CREATE USER/ROLE  (scram 3/10)

### PR DESCRIPTION
**Pull request by knz**
(PR peeled away from #74301; previous PR in series #74842; next PR in series: #74844)
Epic CRDB-5349

This commit introduces the ability to *store* SCRAM hashes.
Note: in this commit they are not yet used to authenticate
SQL clients. Authentication will fail if a SCRAM hash is the only
credential available to a user account.

Release note (sql change): The `CREATE ROLE` and `ALTER ROLE`
statements now accept password hashes computed using the SCRAM-SHA-256
method, for example: `CREATE USER foo WITH PASSWORD
'SCRAM-SHA-256$4096:B5VaT...'`.

As for other types of pre-hashed passwords, this auto-detection can be
disabled by changing the cluster setting
`server.user_login.store_client_pre_hashed_passwords.enabled` to
`false`.

To ascertain whether a SCRAM-SHA-256 password hash will be recognized
as such, orchestration code can use the built-in function
`crdb_internal.check_password_hash_format()` that had been previously introduced.

The encoding of the SCRAM-SHA-256 password should be performed as
follows:

1. take the cleartext password string.
2. generate a salt, iteration count, stored key and server key as per
   [RFC 5802](datatracker.ietf.org/doc/html/rfc5802).
3. encode the hash into the format recognized by CockroachDB: the
   string `SCRAM-SHA-256$`, followed by the iteration count,
   followed by `:`, followed by the base64-encoded salt, followed
   by `$`, followed by the base-64 stored key, followed by `:`,
   followed by the base-64 server key.

### CODE REVIEW

**Comment by knz** [comment]
Note to self: add a check that the iter count is greater than 4096.

**Review by rafiss — APPROVED** [review]
lgtm; i left a comment which maybe can be considered for a separate PR (or if you feel it's easy to add to this one feel free)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74843#-:-MtP7PKJ4S-R5vn1vk3t:b-jhh8j7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)

___
*[-- commits, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74843#-MtP6oLb22ToHutSERYQ):*
instead of linking to the RFC (whose instructions i find hard to follow), what are your thoughts on introducing a builtin function that does these steps? maybe `crdb_internal.generate_scram_hash(salt, iter_count, stored_key, server_key)`

**Comment by knz** [comment]
> what are your thoughts on introducing a builtin function that does these steps? maybe crdb_internal.generate_scram_hash(salt, iter_count, stored_key, server_key)

So you'd want the client to compute the stored_key and server_key, and only use the SQL-side builtin to do the base64 encoding?

**Comment by knz** [comment]
NB: I'm holding off merging this PR until all the PRs in the sequence are ready (otherwise, the entire chain will be auto-closed by github).
